src/internal/storage/fileset/index/reader.go

```diff
@@ -47,30 +47,33 @@ func (r *Reader) Iterate(ctx context.Context, cb func(*Index) error) error {
 	}
 	// Setup top level reader.
 	pbr := r.topLevel()
-	levels := []pbutil.Reader{pbr}
+	levels := []*level{&level{pbr: pbr}}
 	for {
-		pbr := levels[len(levels)-1]
+		pbr := levels[len(levels)-1].pbr
 		idx := &Index{}
 		if err := pbr.Read(idx); err != nil {
-			if errors.Is(err, io.EOF) {
-				return nil
+			if !errors.Is(err, io.EOF) {
+				return errors.EnsureStack(err)
+			}
+			for {
+				if len(levels) == 1 {
```

Comment:
The index tree can be unbalanced after compaction since shards can have different numbers of levels and concatenation only grabs the root indexes. This logic handles the case when the next subtree (shard from a prior compaction) has less levels than the prior. I'm going to look into simplifying this logic further because it is a bit difficult to follow.